### ISSUE TITLE
[Regression] Multiple document_managers fail with 3.5.0

### ISSUE THREAD

**Issue by hjanuschka** [doctrine/DoctrineMongoDBBundle#477]
hello,

first of all thx for this great project we ❤️  it.

using the following yaml:

# 3.4.4

```
parameters:
    env(KRN_HOST_MONGODB_PRIMARY): 'mongo_primary.krn.krone.at'
    env(KRN_HOST_MONGODB_SECONDARY): 'mongo_secondary.krn.krone.at'
    env(MONGODB_DEFAULT_SERVER): "mongodb://%env(KRN_HOST_MONGODB_PRIMARY)%,%env(KRN_HOST_MONGODB_SECONDARY)%:27017"
    env(MONGODB_DEFAULT_CONN): 'krn_default'

doctrine_mongodb:
    default_database: krn_api
    default_connection: '%env(MONGODB_DEFAULT_CONN)%'
    default_document_manager: krn_default

    connections:
        krn_default:
            server: "%env(MONGODB_DEFAULT_SERVER)%"
            options:
                replicaSet: "rs0"

    document_managers:
        krn_default:
            connection: krn_default
            database: krn_api
            #auto_mapping: true
            mappings:
                KRNApiBundle: ~
        krn_lunar_calendar:
            connection: krn_default
            database: krn_lunar_calendar
            mappings:
                KRNApiBundle: { type: annotation, prefix: KRNApiBundle\Document\LunarCalendar }
        krn_reader_letter:
            connection: krn_default
            database: krn_reader_letter
            mappings:
                KRNApiBundle: { type: annotation, prefix: KRNApiBundle\Document\ReaderLetter }
        krn_daily_horoscope:
            connection: krn_default
            database: krn_daily_horoscope
            mappings:
                KRNApiBundle: { type: annotation, prefix: KRNApiBundle\Document\DailyHoroscope }
```

running:
```
bin/console doctrine:mongodb:mapping:info
```

outputs:
```
Found 4 documents mapped in document manager krn_default:

[OK]   KRNApiBundle\Document\DailyHoroscope\ZodiacDay
[OK]   KRNApiBundle\Document\LunarCalendar\Day
[OK]   KRNApiBundle\Document\LunarCalendar\Zodiac
[OK]   KRNApiBundle\Document\ReaderLetter\Letter

```

# 3.5.0
upgrading to `3.5.0`

results in:

```
Found 1 documents mapped in document manager krn_default:

[OK]   KRNApiBundle\Document\DailyHoroscope\ZodiacDay

```

which breaks our application, i tried to dig threw the changelog, but doesnt seem to be any notice of a BC change.

any advices?

thx in advance

**Comment by hjanuschka** [comment]
ok moving 
```
        krn_default:
            connection: krn_default
            database: krn_api
            #auto_mapping: true
            mappings:
                KRNApiBundle: ~
 
```

bottm as the last manager, partially fixes it, output is OK

but still having problems inside the application - finding repositorys

**Comment by hjanuschka** [comment]
```
Exception trace:
 Doctrine\Common\Persistence\Mapping\MappingException::nonExistingClass() at /opt/API/vendor/doctrine/persistence/lib/Doctrine/Common/Persistence/Mapping/RuntimeReflectionService.php:24
 Doctrine\Common\Persistence\Mapping\RuntimeReflectionService->getParentClasses() at /opt/API/vendor/doctrine/persistence/lib/Doctrine/Common/Persistence/Mapping/AbstractClassMetadataFactory.php:251
 Doctrine\Common\Persistence\Mapping\AbstractClassMetadataFactory->getParentClasses() at /opt/API/vendor/doctrine/persistence/lib/Doctrine/Common/Persistence/Mapping/AbstractClassMetadataFactory.php:284
 Doctrine\Common\Persistence\Mapping\AbstractClassMetadataFactory->loadMetadata() at /opt/API/vendor/doctrine/persistence/lib/Doctrine/Common/Persistence/Mapping/AbstractClassMetadataFactory.php:183
 Doctrine\Common\Persistence\Mapping\AbstractClassMetadataFactory->getMetadataFor() at /opt/API/vendor/doctrine/mongodb-odm/lib/Doctrine/ODM/MongoDB/DocumentManager.php:287
 Doctrine\ODM\MongoDB\DocumentManager->getClassMetadata() at /opt/API/vendor/doctrine/mongodb-odm-bundle/Repository/ContainerRepositoryFactory.php:37
 Doctrine\Bundle\MongoDBBundle\Repository\ContainerRepositoryFactory->getRepository() at /opt/API/vendor/doctrine/mongodb-odm/lib/Doctrine/ODM/MongoDB/DocumentManager.php:524
 Doctrine\ODM\MongoDB\DocumentManager->getRepository() at /opt/API/src/KRNApiBundle/Services/MongoDocumentApi.php:15
 KRNApiBundle\Services\MongoDocumentApi->getMongoRepository() at /opt/API/src/KRNApiBundle/Services/MongoDocumentApi.php:36
 KRNApiBundle\Services\MongoDocumentApi->findMulti() at /opt/API/src/KRNApiBundle/Services/MongoDocumentApi.php:24
 KRNApiBundle\Services\MongoDocumentApi->find() at n/a:n/a
 call_user_func_array() at /opt/API/src/KRNApiBundle/Services/AbstractService.php:157
 KRNApiBundle\Services\AbstractService->getUncachedData() at /opt/API/src/KRNApiBundle/Services/MongoDocumentData.php:62
 KRNApiBundle\Services\MongoDocumentData->deleteDocuments() at /opt/API/src/KRNApiBundle/Services/Extern/ReaderLetter/LetterItem.php:115
 KRNApiBundle\Services\Extern\ReaderLetter\LetterItem->deleteItemsByDate() at /opt/API/src/KRNApiBundle/Services/manager/ReaderLetterManager.php:55
 KRNApiBundle\Services\manager\ReaderLetterManager->deleteReaderLetterItemsByDate() at /opt/API/src/KRNApiBundle/Services/Extern/ReaderLetter/ReaderLetterImporter.php:120
 KRNApiBundle\Services\Extern\ReaderLetter\ReaderLetterImporter->import() at /opt/API/src/KRNApiBundle/Services/Extern/ReaderLetter/ReaderLetterImporter.php:37
 KRNApiBundle\Services\Extern\ReaderLetter\ReaderLetterImporter->importByCommand() at /opt/API/src/KRNApiBundle/Command/ReaderLetter/ImportDayCommand.php:86
 KRNApiBundle\Command\ReaderLetter\ImportDayCommand->execute() at /opt/API/vendor/symfony/console/Command/Command.php:251
 Symfony\Component\Console\Command\Command->run() at /opt/API/vendor/symfony/console/Application.php:904
 Symfony\Component\Console\Application->doRunCommand() at /opt/API/vendor/symfony/framework-bundle/Console/Application.php:89
 Symfony\Bundle\FrameworkBundle\Console\Application->doRunCommand() at /opt/API/vendor/symfony/console/Application.php:262
 Symfony\Component\Console\Application->doRun() at /opt/API/vendor/symfony/framework-bundle/Console/Application.php:75
 Symfony\Bundle\FrameworkBundle\Console\Application->doRun() at /opt/API/vendor/symfony/console/Application.php:145
 Symfony\Component\Console\Application->run() at /opt/API/bin/console:39

krn:reader-letter:import-day [-s|--source-date SOURCE-DATE] [-z|--target-date [TARGET-DATE]] [-o|--allow-overwrite [ALLOW-OVERWRITE]] [-h|--help] [-q|--quiet] [-v|vv|vvv|--verbose] [-V|--version] [--ansi] [--no-ansi] [-n|--no-interaction] [-e|--env ENV] [--no-debug] [--] <command>
```

**Comment by hjanuschka** [comment]
calling `->getRepository()`  with `KRNApiBundle::Letter`

**Comment by hjanuschka** [comment]
nevermind using `KRNApiBundle:ReaderLetter\\Letter` fixed it

**Comment by alcaeus** [comment]
The first “issue” is a side effect of #471. Previously, the configuration was stored per connection. To fix an issue when using multiple connections this was changed to being stored per document manager. That’s why you suddenly only see one document mapped, which would have been the correct configuration to begin with.

As for the second issue, that’s on #473 - I just don’t know what exactly is provoking it. I suggest using class names instead of that weird bundle syntax. Thanks to the ::class constant you’ll have a proper identifier instead of a magic string.

**Comment by hjanuschka** [comment]
@alcaeus thx for reply, the above fixes did'nt fix it 100% - i also had to remove my various doc-managers


and only kept:

```
    document_managers:
        krn_default:
            connection: krn_default
            database: krn_api
            #auto_mapping: true
            mappings:
                KRNApiBundle: ~

```

otherwise it ended up anyway in the default database.
so i removed them, and set the `db=`  in the `@Document`  annotation.

**Comment by alcaeus** [comment]
That shouldn’t be necessary. I’ll check it out tomorrow.

**Comment by hjanuschka** [comment]
thx, i am happy anyway, less config :)


so what i needed was:

  - get rid of Bundle:Document notation using `KRNApiBundle:Namespace\\Document`
  - remove document managers - as only the default one was used anyway
   - for using diff. dbs switched to `@Document(db='xxxx')` annotation in the documents itself.

> and we are happily running 3.5.0 in production under heavy load 👏 

if you need me to test something, just let me know. and keep up the good work ❤️

**Comment by alcaeus** [comment]
So, the reason why I wanted to take a closer look was because I wasn't sure how this worked in the first place.

Turns out, the reason it worked was because you'd have a separate configuration object for each document manager, which allows you to have different mapping configuration with the same alias for each document manager (as you did by only assigning a part of the bundles' documents to each document manager). Since one of the document managers has the same name as the connection (`krn_default`), the issue fixed in #471 does not appear. However, the fix causes every document manager to share the connection configuration for `krn_default`, which prevents re-using mapping aliases and leaves you with a single mapping alias for the last document manager in the configuration.

This is a difficult issue to deal with: for one, it's a BC break, however the configuration above only works because of the shared name between one document manager and one connection. Renaming the connection to `krn_default_connection` will break the entire container with an invalid configuration service reference. In the end, I believe it's more feasible to use a single document manager for all documents - the fact that they're in the same bundle makes them related, so in my opinion they shouldn't have been split up in the first place. As you've mentioned, they can still be stored in separate databases by setting the `db` option in the document mapping; however, using a single document manager allows you to handle relationships between the documents.

However, this is all moot with the upcoming 4.0 release: since the new ODM version no longer needs to pass a configuration object to the underlying connection, the configuration object will be owned by the document manager again starting with 4.0, which makes the configuration mentioned above valid again, without the drawback of needing a document manager with the same name as the connection.

One last thing, if I may:
> get rid of `Bundle:Document` notation, instead using `KRNApiBundle:Namespace\\Document`

Allow me to suggest getting rid of your replacement as well. Using new annotations libraries give you support for class constants, so you can safely do this:
```
@ODM\Document(repositoryClass=MyRepository::class)
```
as well as this:

```
$this->documentManager->getRepository(MyDocument::class);
```

The big advantage is that the class names are now identifiers that your IDE can search for, instead of magic strings with a meaning that only Symfony understands.

Given that the configuration above was always somewhat broken and you've found a good workaround, I'm closing this issue as fixing this issue isn't really worth the trouble. I may reconsider this if more people run into this issue.

**Comment by hjanuschka** [comment]
thx for the great Explantation, and all the effort you put in the research.

i am going to follow your advice and change the symfony magic identifier to the `::class` one.

i don't really know why we wen down the "multi-manager" way, i am more than happy with the current behaviour, as it simplifies the config alot.

thank you very much 🎉
